proposal/package-importer.d.ts.md

```diff
@@ -0,0 +1,386 @@
+- Let `styleValue` be the value of `style` in `packageManifest`.
+- If `styleValue` is a relative path with an extension of `css`, return the
+  `packagePath` appended with `styleValue`.
+- Otherwise return the result of [resolving a file url] with `packagePath`.
+
+[resolving the root directory for a package]:
+    #resolving-the-root-directory-for-a-package
+[resolving a file url]: ../spec/modules.md#resolving-a-file-url
+
+### Resolving a package name
+
+This algorithm takes a string, `url`, and returns the portion that identifies
+the Node package.
+
+- If `url` starts with `@`, it is a scoped package. Return the first 2 [url path
+  segments], including the separating `/`
+- Otherwise, return the first url path segment.
+
+[url path segments]: https://url.spec.whatwg.org/#url-path-segment
+
+### Resolving the root directory for a package
+
+This algorithm takes a string, `packageName`, an absolute URL
+`currentDirectory`, and an optional absolute URL `previousUrl`, and returns an
+absolute URL to the root directory for the most proximate installed
+`packageName`.
+
+> We need to replicate Node's behavior, as defined in [Loading from node_modules
+> folders], of walking up the directory chain to find packages from
+> `previousUrl`, so that packages can use the correct version of the packages
+> that they depend on.
+
+- If `previousUrl` is null, return the result of `require.resolve` with the
+  `packageName`.
+- While `rootDirectory` is undefined:
+  - Remove the final path segment from `previousUrl`
+  - If `node_modules` is the new final path segment of `previousUrl`, continue.
+  - Let `potentialPath` be `previousUrl` appended with `node_modules/` and
+    `packageName`.
+  - If `potentialPath` is a directory, let `rootDirectory` be `potentialPath`.
+  - Otherwise, if `previousUrl` is the root of the file system, throw an error.
+- Return `rootDirectory`.
+
+[loading from node_modules folders]:
+    https://nodejs.org/api/modules.html#loading-from-node_modules-folders
+
+### Resolving package exports
+
+This algorithm takes a string `condition`, a package.json value
+`packageManifest`, and a `pkg` URL `url`. It returns a file URL or null.
+
+This algorithm should follow the Node resolution algorithm, as defined in the
+Node documentation under [Node Modules] and [Conditional Exports]. Where
+possible in Node, it can use [resolve.exports] which exposes the Node
+resolution algorithm, allowing for per-path custom conditions, and without
+needing filesystem access.
+
+[resolve.exports]: https://github.com/lukeed/resolve.exports
+[Conditional Exports]: https://nodejs.org/api/packages.html#conditional-exports
+[Node Modules]: https://nodejs.org/api/modules.html#all-together
+
+## Ecosystem Notes
+
+The new `usePkgImporter` option will not be available in the [Legacy JS API].
+Third-party applications that don't support the modern API will be unable to use
+the built-in package importer. Some notable examples follow.
```

Comment:
Given that so many existing plugins still don't support the modern API (:skull:), I wonder if we _should_ support the Node.js package importer for the legacy API. It might not be so hard: we could just say that if `usePkgImporter` is set, `pkg:` loads go through the `pkg:` importer and if that returns null we invoke the legacy importer logic.